Fixes for icap-client

nastya opened the pull request.

nastya committed:
Temporary fix of reading bug

nastya committed:
Change path in go.mod

nastya committed:
Change includes to use this library instead of original

nastya committed:
Rewrite parsing of server responses

Use information in `Encapsulated` header to find
end of message properly.

nastya committed:
Add test for `transport`

Add test checking that ICAP server responses
can be read properly in `transport`.

nastya committed:
Remove attempts to stop test ICAP server

`github.com/egirna/icap` does not support shutdown of ICAP server.
So it is actually not possible to stop already started servers
in runtime.
The removed code did not really stop the ICAP server,
it was just useless complication.

nastya committed:
Fix `SetPreview`

Fix bug in processing 0-length bodies

nastya committed:
Fix `bodyEndIndicator`

Message must end with double CRLF.
Single ending CRLF will lead to error in client:DoRemaining.

nastya committed:
Remove unused code

nastya committed:
Make `addFullBodyInPreviewIndicator` more reliable

nastya committed:
Enhance processing of CRLF-suffixes for requests

nastya committed:
Fix test for RESPMOD

Use `res-hdr` instead of `req-hdr` as far as
the test concerns only response.

nastya committed:
Return error for unsupported protocol versions

Current version supports only ICAP/1.0 and HTTP/1.1 so
in case of other protocols obtained while parsing ICAP response
we should return error.

nastya committed:
Fix some typos and add notes

Add `logDebug` message in case we encountered OPTIONS request with body

nastya committed:
Add option for using absolute URL

Make using absolute URL in embedded HTTP request optional.
Also change the way the request line is formed in case of using
absolute URL (previous implementation used assumption that
`url.EscapedPath()` is contained in `httputil.DumpRequestOut()`
request line).

nastya committed:
Send only request header in RESPMOD

nastya committed:
Filter hop-by-hop headers

nastya committed:
Rename library module name

Use `github.com/solidwall/icap-client` instead of
`github.com/nastya/icap-client`.

Invizory commented on transport.go:
```diff
@@ -95,26 +163,20 @@ func (t *transport) read() (string, error) {
-		if strings.HasSuffix(string(data), "0\r\n\r\n") {
-			logDebug("End of the file detected by 0 Double CRLF indicator")
-			break
+		expectedLength := findLastSectionStart(tmp) // this is the beginning of last section in file
```
А для чего это нужно и почему мы берем максимальное из чисел?

nastya replied:
Тут мы пытаемся понять, сколько данных ещё ждать из сокета. Я решила ориентироваться на offset-ы начал секций в пакете (их там может быть несколько, и они разделены double crlf-ом), эти offset-ы должны быть перечислены в header-е ICAP-ответа. Соответственное, максимальное из чисел берём, потому что хотим убедиться, что последняя секция уже прочитана.

Меня тут смущает разве что то, что по идее достаточно одного вызова этой функции `findLastSectionStart`, а у меня она в цикле вызывается. Не понимаю точно сейчас, почему я так сделала, вполне возможно так было удобнее писать код просто, а может я чего-то не учитываю сейчас.

Invizory replied:
Да, я тоже не очень тут понимаю логику. Я понял бы еще, если бы в цикле делалось `findLastSectionStart(data)`, но тут это вызывается от `tmp`, который есть очередной прочитанный кусок из сокета, а в другом куске уже не будет заголовка `Encapsulated`, поэтому оно `0` вернет.

nastya replied:
Так, тут на самом деле был баг в функции `findLastSectionStart`, она не учитывала последний из оффсетов в хедере и из-за этого часто выдавала 0. У меня такое ощущение, что я потеряла часть изменений в этом файле transport.go, потому что уж больно странно такая ошибка выглядит.

По поводу самого вот этого момента, который тебя смутил, тут не получается вне цикла вычислить `expectedLength` (читай, начало последней секции в ответе), потому что возможно такое, что ICAP-сервер маленькими кусочками отдает данные в сокет и может быть нужно несколько итераций пройти, пока получим строку с `Encapsulated: ...`. И да, безусловно нужно передавать всю считанную на текущий момент `data`, а не очередной кусочек `tmp` в эту функцию, это у меня ещё бага была.

nastya replied:
Я поправила `findLastSectionStart` и `read()` и добавила тесты на `findLastSectionStart` отдельно и тест, в котором сервер разбивает ответ на порции ровно по границе секций (то есть нам обязательно знать, что не все секции ещё прочитаны, ориентируясь на данные об оффсетах начал секций, чтобы всё вычитать). Ещё собственно, сделала сервер в тестах медленным и отдающим данные маленькими порциями.

До вот этих последних изменений всё работало по сути так, что читало данные из сокета, пока в хвосте прочитанного не оказывался DoubleCRLF. И поскольку на практике ICAP-сервер сразу весь ответ в сокет загонял (и в юнит-тестах моих тоже), ошибка не была замечена.

nastya replied:
Логика чтения ответа из сокета сейчас такая:

Сам корректный ICAP-ответ - это одна или несколько секций, разбитых по DoubleCRLF. Если в ответе несколько секций, то обязательно должен быть хедер `Encapsulated` в первой секции с указанием начал оффсетов остальных.
Соответственно в цикле мы читаем из сокета, пытаемся узнать оффсет начала последней секции (назвали его `expectedLength` при этом). Если прочитали уже дальше начала этой последней секции и видим, что в конце считанного DoubleCRLF, то останавливаемся. Если у нас была одна секция всего (например, ответ был типа `ICAP/1.0 100 Continue`), то это условие тоже подходит, ожидаемая длина так и останется -1, и дожидаемся DoubleCRLF, после чего останавливается чтение.

А если вдруг вместо вот такого правильного формата сервер пришлёт что-то странное без DoubleCRLF, то цикл остановится по readTimeout с ошибкой.

nastya replied:
@inviz Перепроверь, не видишь ли изъянов в текущей логике считывания ответа из сокета.

А я посмотрю ещё в целом на код либы, потому что правда сложилось такое впечатление, что часть изменений могла потеряться в процессе моих множественных итераций рефакторинга.

Invizory replied:
А если очередное чтение из сокета вернет данные, которые порезаны как раз по содержимому заголовка `Encapsulated`? Тогда длина, которую вычислит эта функция, будет меньше фактической, либо будет равна 0. Такое же может быть?

nastya replied:
Да, думаю ты прав, может. В `findLastSectionStart` стоит ещё проверять, что мы дошли до `\r` и возвращать максимальное из считанных чисел только в этом случае. А если нет, то возвращать -1 (что, соответственно, заставит на следующей итерации пересчитать `expectedLength` уже на считанном полностью заголовке `Encapsulated`, ну или так и будем дальше считывать по кусочкам, пока на очередной итерации полностью не считается, и тогда вернется из `findLastSectionStart` не -1).

nastya replied:
Поправила это в последнем коммите и сделала так, чтобы в юнит-тестах сервер разбивал ответ на ещё меньшие кусочки, чтобы такая ситуация как раз воспроизводилась.

nastya committed:
Enhance tests for transport.go

1) Make test ICAP server stub work slower and send response by portions
to imitate worst case scenario of socket interaction
2) Add test for reading all sections of response artificially dividing
ICAP response by `DoubleCRLF`
3) Add tests for `findLastSectionStart` function

nastya committed:
Fix `findLastSectionStart`

nastya committed:
Fix `transport:read()`

Call `findLastSectionStart` only unless `Encapsulated` header
is found.

nastya committed:
Use go 1.21

nastya committed:
Fix processing of Encapsulated header value

Fix processing of server response when Encapsulated header value is separated
in received data portions.
Return -1 in `findLastSectionStart` in case given input
does not contain full Encapsulated header value.

nastya committed:
Fix trimming CRLF

`strings.Trim(str, CRLF)` removes all trailing and leading `\r` and `\n`
independently from their order.
In this code we need another functionality: trimming all trailing CRLFs.

Inviz commented:
С удовольствем перед сном читаю ваш прогресс. Можете добавлять меня в другие PR, могу быть ревьюером

nastya commented:
> С удовольствем перед сном читаю ваш прогресс. Можете добавлять меня в другие PR, могу быть ревьюером

Я нечаянно )
Даже не догадалась проверить прежде чем mention-ить, что у Артура на гитхабе тот же ник.

nastya commented:
@Invizory @romko11l я просмотрела код, ничего критичного не нашла, добавила одно маленькое исправление ещё в последнем коммите. Если у вас больше нет замечаний, можете мержить эту либу.

Invizory commented:
Я уже одобрил, про Рому не знаю, но если ему тоже ок, то можешь сама мержить
